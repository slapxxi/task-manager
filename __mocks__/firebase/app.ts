import mock from '.';

export default mock;
