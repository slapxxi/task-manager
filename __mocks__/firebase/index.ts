import { set } from 'lodash';

Date.now = () => 69;

let databaseMock;

const snapshot = {
  val() {
    return databaseMock;
  },
  tasks: {
    val() {
      return databaseMock.tasks;
    },
  },
  tags: {
    val() {
      return databaseMock.tags;
    },
  },
};

const mock = {
  __reset() {
    databaseMock = {
      tasks: {
        'first-task': {
          title: 'first',
          createdAt: 1,
        },
        'second-task': {
          title: 'second',
          createdAt: 2,
        },
        'third-task': {
          title: 'third',
          tags: ['html'],
          project: 'test',
        },
      },
      tags: {
        html: {
          name: 'HTML',
        },
        css: {
          name: 'CSS',
        },
      },
      projects: {
        test: {
          name: 'Test Project',
        },
      },
    };
  },
  database() {
    return {
      ref(path: string) {
        if (path === '/') {
          return {
            on(event, fn) {
              fn(snapshot);
              const interval = setInterval(() => {
                fn(snapshot);
              }, 1);
              return interval;
            },
            off(event, fn) {
              clearInterval(fn);
            },
          };
        }
        return {
          on(event, fn) {
            fn(snapshot.tasks);
            setInterval(() => {
              fn(snapshot.tasks);
            }, 1);
          },
          once(event) {
            return new Promise((resolve) => {
              resolve(snapshot.tags);
            });
          },
          set(value) {
            const objectPath = path.replace(/\//g, '.').slice(1);
            set(databaseMock, objectPath, value);
          },
        };
      },
    };
  },
};

export default mock;
