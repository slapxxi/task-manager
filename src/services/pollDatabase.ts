import { createTask, fromEpochTime } from '@lib';
import { DBEntry, DBProject, DBTask, Project, StoreTag, Tag, Task } from '@local/types';
import firebase from 'firebase/app';
import compact from 'lodash-es/compact';
import map from 'lodash-es/map';
import sortBy from 'lodash-es/sortBy';

interface Response {
  tags: Tag[];
  tasks: Task[];
  projects: Project[];
}

function pollDatabase(fn: (state: Response) => void) {
  return firebase
    .database()
    .ref('/')
    .on('value', (snapshot) => {
      const response = snapshot && snapshot.val();
      if (response) {
        const tags = normalizeTags(response.tags);
        const tasks = normalizeTasks(response.tasks, tags);
        const projects = normalizeProjects(response.projects, tasks);
        return fn({ tasks, tags, projects });
      }
      return fn({ tasks: [], tags: [], projects: [] });
    });
}

function normalizeTasks(tasks: DBEntry<DBTask>, tags: Tag[]): Task[] {
  return sortBy(
    map(tasks, (task, key) =>
      createTask({
        ...task,
        id: key,
        tags: matchTags(task, tags),
        deadline: task.deadline === 0 ? undefined : fromEpochTime(task.deadline),
      }),
    ),
    'createdAt',
  );
}

function matchTags(task: DBTask, tags: Tag[]): Tag[] {
  if (task.tags) {
    return compact(task.tags.map((id) => tags.filter((it) => it.id === id)[0]));
  }
  return [];
}

function normalizeTags(tags: DBEntry<StoreTag>): Tag[] {
  return map(tags, (item, key) => ({
    id: key,
    name: item.name || 'empty tag',
  }));
}

function normalizeProjects(projects: DBEntry<DBProject>, tasks: Task[]): Project[] {
  return map(projects, (item, key) => ({
    id: key,
    name: item.name,
    tasks: tasks.filter((t) => t.project && t.project === key),
  }));
}

export default pollDatabase;
