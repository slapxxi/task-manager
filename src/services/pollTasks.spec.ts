import firebase from 'firebase';
import pollDatabase from './pollTasks';

beforeEach(() => {
  // @ts-ignore
  firebase.__reset();
});

it('polls all available tasks', (done) => {
  pollDatabase(({ tasks }) => {
    expect(tasks).toMatchSnapshot();
    done();
  });
});

it('contains tags', (done) => {
  pollDatabase(({ tasks }) => {
    expect(tasks[0].tags).toEqual([]);
    expect(tasks[2].tags).toEqual([{ id: 'html', name: 'HTML' }]);
    done();
  });
});

it('sorts by `createdAt` field', (done) => {
  pollDatabase(({ tasks }) => {
    expect(tasks[0].createdAt).toEqual(1);
    expect(tasks[1].createdAt).toEqual(2);
    done();
  });
});
