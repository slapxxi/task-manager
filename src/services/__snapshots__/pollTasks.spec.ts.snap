// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`polls all available tasks 1`] = `
Array [
  Object {
    "completed": false,
    "createdAt": 1,
    "description": "",
    "id": "first-task",
    "project": null,
    "tags": Array [],
    "title": "first",
  },
  Object {
    "completed": false,
    "createdAt": 2,
    "description": "",
    "id": "second-task",
    "project": null,
    "tags": Array [],
    "title": "second",
  },
  Object {
    "completed": false,
    "createdAt": 69,
    "description": "",
    "id": "third-task",
    "project": "test",
    "tags": Array [
      Object {
        "id": "html",
        "name": "HTML",
      },
    ],
    "title": "third",
  },
]
`;
